### ISSUE TITLE
test(metalgemm): observe graph Q8 codes and scales after successful completion

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12914]
Parent: #12349
Coordinates with: #12853

## Parent context
#12349 strict native Metal P4 parity remains RED. #12853 H/D/S diagnostic observes different projection outputs for host quantization through graph (H), device quantization through graph (D), and original serial native group (S). This child observes the actual device codes/scales; it does not repair arithmetic or qualify parent parity.

## Current state
Frozen HDS source manifest SHA256 893ed8726d0b27b256929a0105e85d29688daf0b9cd8437374fe3561251ffe48; physical session58110 exit1. All four normalized input rows agree exactly; H/D/S channel0 bits bf7e14ef/bf7e14f1/bf7e14f0. Device codes/scales have not been read. H versus S proves projection-route drift under fixed host quantization. H versus D alone does not identify code or scale differences.

## For / Problem / Today / Better because
For exact native Metal parity investigation, determine whether identical normalized inputs produce identical quantization codes and scale bits. Today projected output differences cannot identify these intermediate differences. Direct bounded observation makes the next arithmetic candidate falsifiable.

## Why this is next
Host quantization multiplies by reciprocal with away-from-zero rounding; graph quantization divides and uses floor/ceil around half. Actual buffer evidence is needed before choosing a fix.

## Working spine
One explicitly diagnostic terminal read of the existing QuantizedGraphResult, after successful FinishRead and before Free. Preserve the current projection cache map and retrieve its normalized-input entry; do not encode another quantization. Return defensive code/scale copies and separately log 1152 diagnostic bytes.

## Core through-line
Observe existing device intermediates without changing arithmetic, graph submission, or strict acceptance.

## Scope
```routing
lane: model
paths: ["internal/metalgemm/graph.go", "internal/metalgemm/q4k.m", "internal/model/metal_mtp_verify_panel_darwin_test.go"]
expected_steps: 3
```

## Likely files
- internal/metalgemm/graph.go
- internal/metalgemm/q4k.m
- internal/model/metal_mtp_verify_panel_darwin_test.go

## Scope class
Work unit: leaf. One bounded diagnostic bridge across two packages. Obtain admission covering all three paths before source edits.

## Proposed API and safety design (review before editing)
One exported method, DiagnosticReadQ8(r *QuantizedGraphResult) (codes []int8, scales []float32, diagnosticBytes uint64, err error), on ProjectionGraph. No new exported type, buffer pointer, generic buffer reader, or production caller.

Go rejects nil/freed graph, nil pointer, unsuccessful terminal Finish, foreign/nil result, nonpositive or non-32-divisible elems, and lengths exceeding C int or safe allocation bounds. Add a private finishSucceeded field set only when mg_graph_finish returns success. A completed command buffer with injected post-submit failure remains rejected. Read is valid only after successful Finish; the model witness additionally requires successful FinishRead before calling it. Existing Read/FinishRead behavior is outside scope.

Native bridge independently requires a graph success marker set only for completed non-injected mg_graph_finish, live graph, nonnull destinations, exact element/scale lengths, and a registered quantization pair. At mg_graph_quantize_q8 creation, record private graph-owned metadata associating q, d, elems and their code-int8/scale-float32 roles. Membership in the generic results array alone is insufficient: verify the exact pair, element count, actual buffer byte lengths and Shared storage before memcpy. Validate everything before either copy; return no partial result on error. Metadata is released with the graph. No commit, encoder, wait, or production readback counter increment occurs in the diagnostic bridge.

Copies are 1024 code bytes + 32*4 scale bytes = 1152 for P4/H256. Returned diagnosticBytes is derived from successful validated copies, not a fixture constant. The test logs diagnostic_readbacks=1 and diagnostic_bytes=1152 separately; production receipt remains 4 encoders, 1 packed readback, 20480 read bytes and 6272 upload bytes. Total observed host transfer is explicitly 21632 read bytes including the diagnostic. Returned slices own their memory and remain valid after graph Free. Repeated calls, if made, each perform and report their own diagnostic transfer; no hidden cache.

## Definition of done
- [ ] Independent API/ownership/accounting review accepts this exact source-only design before editing.
- [ ] Candidate modifies only the three admitted files and receives source review.
- [ ] Explicit serialized HIL grant produces actual code/scale comparison and separate diagnostic accounting.
- [ ] Independent review records the observed result and integration disposition; original strict parity stays authoritative.

## Gold-plating boundary
No new quantization algorithm, shader arithmetic, tolerance, oracle edit, broader exported API, production use, engine substitution, additional GPU graph, or performance claim. Preserve frozen HDS source until design acceptance. No HIL authorization is implied by this ticket.

## Acceptance gate
Existing TestMetalMTPP4StrictSerialParity fixture: P4/H256, original Q8 layer0 weights, all four normalized rows bitwise equal, H/D/S native routes unchanged. Compare all 1024 signed codes exactly and all 32 scales by Float32bits. Report hashes, mismatch counts and first differing row/block/lane and host/device values/bits. Neither equality nor inequality is presumed. Preserve every downstream strict assertion and report its RED separately from diagnostic success. Require one test run and zero skips. In the same witness add bounded rejection assertions for pre-Finish access, foreign owner and post-Free access without adding GPU work; validate unconstructible/private type/length and injected-failure rejection through source review within this three-file scope.

## Done condition
Diagnostic delivers witnessed direct quantization evidence with independently accepted lifecycle and accounting boundaries. No parent closure or numerical fix claim.

## Witness
env -u FAK_METAL_MPS CGO_ENABLED=1 GOMAXPROCS=2 GOFLAGS=-p=1 go test ./internal/model -run '^TestMetalMTPP4StrictSerialParity$' -count=1 -timeout=60s -json
Capture before/after full source hashes, native device, JSONL, stderr, separate diagnostic receipt and strict exit. Use existing isolated cache, outer timeout300 seconds.

## Budget
Read-only design first; no source edits before independent acceptance. Subsequently one authorized physical run, outer300/test60, same workload and four graph encoders; 1152 additional CPU readback bytes and bounded metadata/CPU comparisons. No extra GPU dispatches or waits. Coordinator CPU/HIL baton required.

## Duplicate search
Open and closed public issues searched for quantized readback graph. #9267 covers graph ownership, #12853 observes projected outputs; no exact direct code/scale diagnostic child found.

## Problem frame
Centrality: Core
Portfolio tier: 2 serving
P1: advanced - isolate exact native arithmetic discrepancy.
P2: preserved - no performance promotion.
P3: preserved - bounded three-file diagnostic API.
P4: advanced - direct physical intermediate evidence.

## Work estimate
Estimate: 1 point.

## Overall completion contribution
Contribution: 1/4 points within #12349; planning only.

## Completion standard
experiment

## Closure binding
Explicit-path witnessed commit cites child and (fak model), or record a reviewed non-landing experiment disposition. #12349 remains open for complete original acceptance.

**Comment by anthony-chaudhary** [comment]
Source-only candidate prepared after independent design acceptance; no build, test, HIL, commit or landing yet.

- Managed tree: `fak-worker-wt-model-165612fe8f4e`; artifacts `_scratch/issue-12914-source/{candidate.patch,source.sha256,receipt.json}`.
- Exact authored paths: `internal/metalgemm/graph.go`, `internal/metalgemm/q4k.m`, `internal/model/metal_mtp_verify_panel_darwin_test.go`.
- Patch SHA256 `7eac3dd0b35114b44ff04fa4383d30ff02254c228054ab38aa07240784056b7e`; full source manifest `aa7283af34c9b53a40c08c39ebae4311672c9b956889c38af1fee13b95b63313`.
- Diagnostic uses the existing cached Q8 result; Go/native successful-completion checks, registered exact code/scale/element pair, ownership, bounds, actual lengths and Shared storage validate before copying. Native metadata clears before graph buffers release.
- Same fixture adds pre-Finish, foreign Go owner and post-Free rejection checks without added GPU dispatch. Foreign owner is a copied terminal Go handle used only for rejection; original alone owns/frees native storage. Native malformed-private-pair and injected-failure paths remain explicit source-review boundaries in this three-file scope.
- Copies survive Free; compare all 1024 signed codes and 32 scale bit patterns. Separate 1152-byte diagnostic transfer leaves original 20480-read/6272-upload/4-encoder production receipt intact. All H/D/S and downstream strict assertions retained.
- Formatting and diff checks pass. Original HDS sources remain frozen. Lease released.

Next: independent review of this exact source candidate, followed by a separately granted serialized physical witness. Parent #12349 remains RED pending original exact acceptance.

**Comment by anthony-chaudhary** [comment]
One granted physical witness completed on fak-native Apple M3 Pro: executor session56092, PID/PGID1366, exit1; one test run, zero skips. Test0.15s/package0.624s; supervised build/run41.738s. Owned process group and observed descendants absent at terminal; no signals needed; lane released. All1702 frozen source entries unchanged.

Direct quantization observations:
- All1024 signed codes match exactly. Both SHA256 `cdfecda26215f8bef90149c1afb481689f2723facdbbeb7237a0f2f592e7e588`.
- Three of32 scale bit patterns differ. First: row1/token5/block5, host `3cd63ec4`, device `3cd63ec3`.
- Host scale SHA256 `00f8cb88a3a83b03c56be3fffbd306054c226dc350491462a33381bd081e04fb`; device `fd4a30991eedf5828a060482316e13e6850b38193557d5412667579a8864b2e7`.
- The fixture compares every code/scale but persists hashes, counts and first differing location/bits; the other two scale locations and complete raw arrays were not logged.
- Pre-Finish, foreign-owner and post-Free reads rejected; owned copies unchanged after Free. Separate diagnostic read1152bytes; production receipt remains one read20480bytes, upload6272bytes, four encoders. Total observed read21632bytes.

All four normalized rows remain bitwise exact. H/D/S hashes and their pairwise differences remain unchanged. Original strict overall parity remains RED for projection/logits/GDN/KV/hidden/continuation; no tolerance or oracle changed.

Receipt: managed `fak-worker-wt-model-165612fe8f4e/_scratch/issue-12914-hil/receipt.json`, SHA256 `1a33c28ca1e0cdda0fef5709d11b2d64a10490968d0548f2b6008b928e59c598`. Raw `test.jsonl` SHA256 `6070ac396d89d9e124ed0cedfe7a48c74f155764367b9fac80cb5f341b5ce3b8`; adjacent test.log, supervision.json and before/after source checks.

Next: independently review these physical observations, then scope scale-formation investigation separately from the already-observed H-versus-S projection-route drift. No sole-cause inference, parent closure, performance claim or additional run is authorized by this result.
